File: Algorithm/Sort/mergeSort.cpp
#include <vector>
#include <iostream>

using namespace std;

void print(vector<int> &arr) {
	for (auto a : arr) {
		cout << a << " ";
	}
	cout << endl;
}

class MergeSort {
	private:
		static void merge(vector<int> &arr, int left, int middle, int right)
		{
			vector<int> sorted;
			int i, j;

			for (i = left, j = middle; i < middle && j < right;) {
				if (arr[i] > arr[j]) {
					sorted.push_back(arr[j]);
					j++;
				} else {
					sorted.push_back(arr[i]);
					i++;
				}
			}

			while (i < middle) {
				sorted.push_back(arr[i]);
				i++;
			}

			while (j < right) {
				sorted.push_back(arr[j]);
				j++;
			}

			for (i = left; i < right; i++) {
				arr[i] = sorted[i - left];
			}
		}

		static void split(vector<int> &arr, int left, int right)
		{
			if (right - left < 2) {
				return;
			}

			int middle = (left + right) / 2;
			split(arr, left, middle);
			split(arr, middle, right);

			merge(arr, left, middle, right);
		}
	public:
		static void mergeSort(vector<int> &arr)
		{
			split(arr, 0, arr.size());
		}
};

int main(int argc, char const *argv[])
{
	vector<int> vec = {3, 4, 7, 2, 1, 10, 9};
	MergeSort::mergeSort(vec);

	print(vec);
	return 0;
}

File: Algorithm/gosort/merge_sort.go
package main

type MergeSort struct {
	items    []int
	isSorted bool
}

func NewMergeSort(items []int, algo SortAlgorithm) *MergeSort {
	return &MergeSort{
		items: items,
	}
}

func (m *MergeSort) Perform() []int {
	if m.isSorted {
		return m.items
	}

	m.split(0, len(m.items))
	m.isSorted = true

	return m.items
}

func (m *MergeSort) split(left int, right int) {
	// Basis of recursion:
	// If only 1 element left
	if right-left < 2 {
		return
	}

	middle := (right + left) / 2
	m.split(left, middle)
	m.split(middle, right)

	m.merge(left, middle, right)
}

func (m *MergeSort) merge(left, middle, right int) {
	j := middle
	i := left
	s := []int{}

	for i < middle && j < right {
		if m.items[i] < m.items[j] {
			s = append(s, m.items[i])
			i++
		} else {
			s = append(s, m.items[j])
			j++
		}
	}

	for i < middle {
		s = append(s, m.items[i])
		i++
	}

	for j < right {
		s = append(s, m.items[j])
		j++
	}

	for i = left; i < right; i++ {
		m.items[i] = s[i-left]
	}
}
